### DISCUSSION TITLE
Can't get audio over HDMI on rpi

### DISCUSSION THREAD

**Discussion by DockDuckDaisy** [mikebrady/shairport-sync#1847] [Q&A]
I want to run shairport-sync on an rpi 3B and output the audio over HDMI.
I have tried for so long to solve this, but I am at the end of my rope. Can anyone help?

My RPi 3B has been running shairport-sync for many years. At one point something broke in an update and shairport-sync became unstable. It would play audio but crash after a short time whenever I tried to use it. I did a clean install of the OS, switching to a lite version with no desktop, and ALSA as the backend. Shairport-sync works fine over the headphone jack, but when I switch to HDMI it refuses to work. Speakertest and mpv both successfully output over HDMI, so it is not a problem with the hardware or signal chain.

Shairport-sync over HDMI starts and runs, but as soon as I connect from a source device, the process crashes.

Verbose output of the failure:
```
0.006417031 "audio_alsa.c:551" *fatal error: audio_alsa: Could not automatically set the output format for device "hw:vc4hdmi": Invalid argument
0.000221927 "rtsp.c:531" Connection 1: cancelled.
101.700378711 "shairport.c:1728" emergency exit
```

sps-alsa-explorer tells me:
```
Device Full Name:    "hw:vc4hdmi"
Short Name:          "hw:1"
Shairport Sync can not use this device.
```
This seems significant... but in all my searching, I only find dead ends.

shairport-sync -V
```
4.3.2-2-g165431a8-AirPlay2-smi10-OpenSSL-Avahi-ALSA-soxr-sysconfdir:/etc
```

Other details:
Raspberry Pi 3B
Raspberry Pi OS (Legacy) Lite - Bullseye
HDMI
ALSA

**Comment by mikebrady** [comment]
Thanks for the post. Would [this](https://github.com/mikebrady/shairport-sync/discussions/1499) discussion be relevant at all?

Also, just wondering if the device you plug in to the HDMI port has changed at all?

**Reply by DockDuckDaisy** [reply]
Hi, thank you so much for replying!
My hardware is unchanged, same old devices all the way.

I already found and followed the linked thread. But alas, it is a dead end for me. It mentions "Shairport Sync can not use this device.", which seems like the best clue.

"dtparam=audio=on" will turn on the audio card that powers headphone jack, and load it's driver. The HDMI is unaffected by this as far as I can tell. I am able to produce audio over HDMI using speakertest/mpv with or without that parameter enabled. And the behaviour of shairport-sync is the same (if directed to use the hdmi card).

I have also found and followed the common advice to set these parameters in /boot/config.txt
hdmi_group=1
hdmi_mode=1
hdmi_drive=2
Those seem to be required, but my problem persists.

**Reply by DockDuckDaisy** [reply]
Hi, not sure if you saw my followup reply here?

If HDMI out on Rpi using ALSA is not supported, what should I switch to?

**Comment by mikebrady** [comment]
Apologies, I missed this. Let me come back on this.

**Comment by mikebrady** [comment]
Thanks again for the reminder.

(Some time ago, possibly a year or so, it seems as if drivers were changed -- e.g. see a discussion [here](https://lists.yoctoproject.org/g/yocto/topic/99727959#msg60422) -- and this may be the cause of the difficulty.)

So, just playing about with a Raspberry Pi 3B on the latest version of Raspberry Pi OS.
In this case, the `/boot/config.txt` is untouched: there no changes to the `dtparam=audio=on` and none of those `hdmi_group=...` settings have been added.

```
Raspberry Pi 3 Model B Rev 1.2
Linux RaspberryPi3B 6.6.28+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.6.28-1+rpt1 (2024-04-22) aarch64 GNU/Linux
```
Now, if a HDMI cable is plugged into the Pi and connected at the other end to the output device (e.g. a TV) with is switched on, then by substituting `hdmi` for `hw` in the `alsa` `output_device` setting, it works as an output device. In your case, you would use `output_device = "hdmi:1";` instead of `output_device = "hw:1";`.


If the HDMI cable is disconnected, it stops working. When plugged back in, it takes a few seconds for it to be recognised and it works again.

Would you be kind enough to try that and see if it works, please? You should set the `log_verbosity` to `1` and enable `statistics` (both in the `diagnostics` section of the configuration file).

I'm afraid the `sps-alsa-explore` tool doesn't check for devices with the`hdmi:` prefix at present, but I'll post an update to it soon.

In my case, it reports the the HDMI output a suitable mixer called `"PCM"` but I don't know if that's specific to the device that the HDMI cable is connected to... Might be worth trying...

**Reply by DockDuckDaisy** [reply]
Preliminary result: Yey!

I reinstalled a fresh version of shairport-sync (following the build instructions) and switched to development branch. 
Then I changed the config to enable logging and statistics, as well as output_device = "hdmi:vc4hdmi

It now works! I get audio out through hdmi.

I am running rpi3B--hdmi-->reciever (there is a TV connected, but it is speakerless, and unpowered most of the time) 

Hot unplug and reconnect of hdmi at the rpi3b end does not break anything. Audio resumes after plugging in.

My /boot/config.txt is still modified as before, although I did set dtparam=audio=on before starting the experiment. 

I will let this run for a while to see that it keeps going, but so far so good!
Do you still want logs if everything is working fine?

**Reply by mikebrady** [reply]
That’s great news. In fact, thanks to your prompting, I’ve learned quite a bit.  Working through this, the help section of Shairport Sync has been corrected as well as `sps-alsa-explore`. 

There’s no real need for the logs now, thanks. But if things go wrong, of course…

**Comment by mikebrady** [comment]
So I just pushed an update to [`sps-alsa-explore`](https://github.com/mikebrady/sps-alsa-explore/tree/experimental) in the `experimental` branch. It should detect HDMI devices and mixers properly. If you got a chance to try it and see if it's working for you, it would be appreciated.

**Reply by DockDuckDaisy** [reply]
My git-fu is weak, but I got the experimental sps-alsa-explore set up. It does indeed provide a different result, and seems to be properly detecting the hdmi. I will proceed to check the other posts you made and reply to them as well.

>
Device Full Name:    "hdmi:vc4hdmi"
Short Name:          "hdmi:0"
This device seems suitable for use with Shairport Sync.
Possible mixers:     "PCM",0                 Range:  51.00 dB
The following rate and format would be chosen by Shairport Sync in "auto" mode:
     Rate              Format
     44100             S32_LE

In case it matters, here is the tag list.
1.0
1.0.1
1.0.3d3
1.0.3d4
1.1
1.2
1.2-dev
1.2.1
1.2.1-dev

**Reply by mikebrady** [reply]
You've done well, thanks!

So, it's worth trying Shairport Sync now with the `output_device = "hdmi:vc4hdmi"`.

AFAIK that mixer called PCM doesn't do anything, so best leave the `mixer_control_name` commented out...

**Comment by DockDuckDaisy** [comment]
Thank you! I will test it out and report back.

**Reply by mikebrady** [reply]
Thanks. Just FYI I've noticed that the `development` and `master` versions of Shairport Sync can't access that `PCM` mixer. I'll update the `development` version to be able to see it in the next day or so, but AFAIK it doesn't seem to be a volume control anyway...

**Comment by Tiefflieger** [comment]
Nothing was working to get HDMI Audio Output to work on my raspberry pi 4 with the docker options. So I investigated on system level to see if the HDMI audio output was possible. I made changes to the /boot/config.txt as mentioned here: [Enable HDMI Sound](https://raspberrypi.stackexchange.com/questions/32717/how-to-enable-sound-on-hdmi)

```
# uncomment to force a specific HDMI mode (this will force VGA)
hdmi_group=1
hdmi_mode=1
# uncomment to force HDMI mode rather than DVI. This can make audio work in DMT (computer monitor) modes
hdmi_drive=2
```
I used this command, my HDMI is connected to the second port:
`docker run -d  --restart unless-stopped --net host --device /dev/snd     mikebrady/shairport-sync -a Airpi -- -d hdmi:2`

**Reply by mikebrady** [reply]
Thanks for this!
